### ISSUE TITLE
Build Go Packages as Non-Root User

### ISSUE THREAD

**Issue by bramwelt** [hyperledger-archives/fabric#1790]
## Description

Builds of hyperledger [docker images](https://github.com/hyperledger/fabric/blob/master/Makefile#L124) generate files under `build/docker/pkg` owned as root. These files should be owned by the user running make to ensure they can be properly cleaned up using `make clean`
## Describe How to Reproduce

make unit-test -> produces `build/docker/pkg` owned by root.

**Comment by ghaskins** [comment]
This looks promising: http://man7.org/linux/man-pages/man7/user_namespaces.7.html

**Comment by ghaskins** [comment]
Worst case, we can revert this commit https://github.com/hyperledger/fabric/commit/85837442c2dfb92cf0be0455581387bfbe0c2f48

There will be a performance penalty at build time, but that is better than what we have today.

**Comment by rameshthoomu** [comment]
@ghaskins : Could you please revert the above commit# (https://github.com/hyperledger/fabric/commit/85837442c2dfb92cf0be0455581387bfbe0c2f48) for the short term fix and also to continue with builds on Z and X86. As discussed yesterday in slack channel, we will work on running docker containers in non-root mode for the long term fix.

**Comment by ghaskins** [comment]
Currently testing this:

```
commit 8ff7b83fb95a552b9bd3777f8ec55f5b9de3548d
Author: Gregory Haskins <gregory.haskins@gmail.com>
Date:   Tue Jun 14 13:39:25 2016 -0400

    Run docker builds as the host user

    Fixes #1790

    Signed-off-by: Greg Haskins <gregory.haskins@gmail.com>

diff --git a/Makefile b/Makefile
index c7a8ee4..4f613c9 100644
--- a/Makefile
+++ b/Makefile
@@ -38,6 +38,8 @@
 PROJECT_NAME=hyperledger/fabric
 PKGNAME = github.com/$(PROJECT_NAME)
 CGO_FLAGS = CGO_CFLAGS=" " CGO_LDFLAGS="-lrocksdb -lstdc++ -lm -lz -lbz2 -lsnappy"
+UID = $(shell id -u)
+DOCKER_USER_NS="host"

 EXECUTABLES = go docker git
 K := $(foreach exec,$(EXECUTABLES),\
@@ -110,9 +112,10 @@ $(GOPATH)/bin/%:
 # This is necessary since we cannot guarantee that binaries built
 # on the host natively will be compatible with the docker env.
 %/bin/protoc-gen-go: build/image/base/.dummy Makefile
-       @echo "Building $@"
+       @echo "Building $@ as UID $(UID)"
        @mkdir -p $(@D)
        @docker run -i \
+               --user=$(UID) --userns=$(DOCKER_USER_NS) \
                -v $(abspath vendor/github.com/golang/protobuf):/opt/gopath/src/github.com/golang/protobuf \
                -v $(abspath $(@D)):/opt/gopath/bin \
                hyperledger/fabric-baseimage go install github.com/golang/protobuf/protoc-gen-go
@@ -128,6 +131,7 @@ build/docker/bin/%: build/image/src/.dummy $(PROJECT_FILES)
        @echo "Building $@"
        @mkdir -p build/docker/bin build/docker/pkg
        @docker run -i \
+               --user=$(UID) --userns=$(DOCKER_USER_NS) \
                -v $(abspath build/docker/bin):/opt/gopath/bin \
                -v $(abspath build/docker/pkg):/opt/gopath/pkg \
                hyperledger/fabric-src go install github.com/hyperledger/fabric/$(TARGET)
diff --git a/scripts/provision/docker.sh b/scripts/provision/docker.sh
index b47b21f..774cf84 100755
--- a/scripts/provision/docker.sh
+++ b/scripts/provision/docker.sh
@@ -97,6 +97,7 @@ do
 done`
 COPY scripts $REMOTESCRIPTS
 RUN $REMOTESCRIPTS/common.sh
+RUN chmod a+rw -R /opt/gopath
 EOF
```
